BCStudentSoftwareDevTeam/celts
# Rework SLC Proposal Forms

Fixes issues #843, #832 and #820 
## Changes
- Reworked FileHandler so that it works with events and courses
- Improved the middle page of the SLC Proposal/made it better match the original document. This includes:
1. changed how isRegularlyOccurring works and added isPreviouslyApproved, previouslyApprovedDescription, and hasSlcComponent fields to the Course model.
2. Adding a place for the user to upload a copy of the course syllabus 
3. Adding a new "Save and Exit" button on the middle page of the SLC proposal and renamed "Save for Later" to "Save and Exit" on the last page
4. Added a "Exit" button on the last page of the SLC proposal when a used the "View" action
- Spaced the SLC questions out from each other on the last page of the proposal
- Added missing text to both the first and last pages of the SLC proposal
- Added a new "Print" option to the action dropdown to let a user download/print their proposal (this option was also added to the Admin "Course Management" page
- Added a text blurb to the Faculty "Course Proposal" page that, among other things, lets the user download a copy of the official proposal form from CELTS
- Combined "courseManagement.js" and "slcManagement.js" files
- Changed the course status message from "INCOMPLETE" to "IN_PROGRESS" and added a hover to them when they appear in the table (the "Status" section) to let the user know what they mean
- Rewrote test to match all the changes we made
- Made user that if a course is renewed through the "Renew" action then "isPreviouslyApproved" is checked by default
- Added test data to make sure all fields on the SLC questionnaire can be edited.

## Testing
- Pull the branch and make sure it is up to date
- Reset your database and run the tests to verify they all pass
- Run the application with "flask run" 
- Verify you are still able to attach a file to an event and that you are also able to delete an event attachment
- Navigate to the "Course Proposals" page
- Verify that you are able to download a copy of the SLC proposal form out of the "here" hyper link in the text blurb
- Hover over the Status of any of the existing proposals and make sure a hover pops up and that its contents match what it is attached to
-  Click the "Create Course Proposal" button and go through the proposal process. 
1. On the middle page of the proposal verify all of the new options work and save to the database as expected
2. Upload a file to the syllabus section and make sure it is actually uploaded
3. Delete the file you have uploaded and make sure it is actually deleted
5. On the middle page use the "Save and Exit" button and verify it has saved the changes made to that middle page when you return to it 
6. Once you finish the proposal test both the "Save and Exit" and "Submit" buttons on the last page and make sure any changes are saved and the status that is displayed on the table matches the action you took (Save for Later the status should be "In Progress" and Submit should have a status of "Submitted")
7. Using the Action drop-down on the course table verify that when you use the "Print" action, a new window is opened with a print screen (don't actually print the form, set the "Destination" as "Save as PDF") and after click "Print" or "Close" inside the print screen that new window is closed and that file is downloaded
8. Using the "View" action from the drop-down, go through all the pages and make sure you are not able to edit any fields and that all the buttons correspond to what they should when a user is viewing a proposal
